### ISSUE TITLE
CentOS 8.2 Graphene Failures

### ISSUE THREAD

**Issue by anjalirx-intel** [oscarlab/graphene#2474]
## Issue-1: Graphene Installation Location in CentOS
While installing Graphene in Ubuntu, graphene is getting installed in /usr/local/bin which is part of default environment path. Whereas in CentOS, the default environment path is /usr/bin but Graphene gets installed in /usr/local/bin.
However if we add the /usr/local/bin path to the environment file, it works properly.
But graphene should be installed at /usr/bin in CentOS

## Issue-2: Graphene Example - Graphene-SGX fails to generate token for python workload

Error:-
```
Traceback (most recent call last):
  File "/usr/local/bin/graphene-sgx-sign", line 5, in <module>
    sys.exit(main())
  File "/usr/local/lib64/python3.6/site-packages/graphenelibos/sgx_sign.py", line 824, in main
    return main_sign(manifest, args)
  File "/usr/local/lib64/python3.6/site-packages/graphenelibos/sgx_sign.py", line 743, in main_sign
    expanded_trusted_files = list(get_trusted_files(manifest).items())
  File "/usr/local/lib64/python3.6/site-packages/graphenelibos/sgx_sign.py", line 183, in get_trusted_files
    path = Path(resolve_uri(val, check_exist))
  File "/usr/local/lib64/python3.6/site-packages/graphenelibos/sgx_sign.py", line 148, in resolve_uri
    raise ManifestError(f'Cannot resolve {uri} or the file does not exist.')
graphenelibos.sgx_sign.ManifestError: Cannot resolve file:/usr/lib64/python3/dist-packages/ or the file does not exist.
```

CentOS:- python3 -m site
```
[intel1@localhost python]$ python3 -m site
sys.path = [
    '/home/intel1/Anjali/graphene/Examples/python',
    '/usr/lib64/python36.zip',
    '/usr/lib64/python3.6',
    '/usr/lib64/python3.6/lib-dynload',
    '/usr/local/lib64/python3.6/site-packages',
    '/usr/lib64/python3.6/site-packages',
    '/usr/lib/python3.6/site-packages',
]
USER_BASE: '/home/intel1/.local' (exists)
USER_SITE: '/home/intel1/.local/lib/python3.6/site-packages' (doesn't exist)
ENABLE_USER_SITE: True

```
After debugging we found that Python disthome which is mounted in the python.manifest.template file using
`fs.mount.pydisthome.path = "{{ python.distlib }}"`
translates in Ubuntu to
`fs.mount.pydisthome.uri = "file:/usr/lib/python3/dist-packages" `
where in CentOS it translates to
`fs.mount.pydisthome.uri = "file:/usr/lib64/python3/dist-packages" `
but CentOS doesn't have any dist-packages folder.

## Steps to reproduce
cd Examples/python
make SGX=1 all
make SGX=1 check

## Expected results
./run-tests.sh > TEST_STDOUT 2> TEST_STDERR

## Actual results
Traceback (most recent call last):
  File "/usr/local/bin/graphene-sgx-sign", line 5, in <module>
    sys.exit(main())
  File "/usr/local/lib64/python3.6/site-packages/graphenelibos/sgx_sign.py", line 824, in main
    return main_sign(manifest, args)
  File "/usr/local/lib64/python3.6/site-packages/graphenelibos/sgx_sign.py", line 743, in main_sign
    expanded_trusted_files = list(get_trusted_files(manifest).items())
  File "/usr/local/lib64/python3.6/site-packages/graphenelibos/sgx_sign.py", line 183, in get_trusted_files
    path = Path(resolve_uri(val, check_exist))
  File "/usr/local/lib64/python3.6/site-packages/graphenelibos/sgx_sign.py", line 148, in resolve_uri
    raise ManifestError(f'Cannot resolve {uri} or the file does not exist.')
graphenelibos.sgx_sign.ManifestError: Cannot resolve file:/usr/lib64/python3/dist-packages/ or the file does not exist.
make: *** [Makefile:40: python.manifest.sgx] Error 1

**Comment by stefanberger** [comment]
Maybe https://github.com/oscarlab/graphene/pull/2319 helps you?

**Comment by anjalirx-intel** [comment]
@stefanberger Yes #2319 helps.

**Comment by dimakuv** [comment]
# Analysis of Issue 1

CentOS/Fedora has a problem with `/usr/local/bin` where we install Graphene (actually, Meson installs Graphene; this is the default in Meson).

The `PATH` envvar in CentOS/Fedora doesn't contain `/usr/local/bin` by default. The rationale seems to be "CentOS/RHEL/Fedora is security-oriented and doesn't want the users to run some malicious graphene-sgx  binary by default" (see e.g. https://security.stackexchange.com/questions/136990/whats-the-motivation-for-excluing-usr-local-bin-from-roots-path).

It also seems that these distros prefer `/usr/bin` for locally installed packages (via proper package install via `yum`). And `/usr/local/bin` is used for locally installed packages built from source (via `make install` / `ninja install`). So Graphene is doing the right thing here.

I tried to check what other apps think about this:
- Meson defaults to `/usr/local` prefix always, for any distro. Meson itself has a note to its own installation: https://github.com/mesonbuild/meson/blob/5eeddd682f9e6df93644030c4faf699f91ba8a45/packaging/macpages/English.lproj/conclusion.html. So Meson is installed under `/usr/local/bin` and then the docs ask to add it to the `PATH`.
- Similarly, HashiCorp also has a note in the docs to add its installation path to the `PATH`.

In other words, when installing from the source, it looks like that at least some popular apps simply say "Please add `/usr/local/bin` to your `PATH` variable if not found".

We should do the same: add some text to our documentation.

**Comment by dimakuv** [comment]
## Analysis of Issue 2

Here is the problem (see also https://github.com/oscarlab/graphene/pull/2319 for full context):

We have our Python/PyTorch examples that rely on system-wide installation of Python. Our example manifest file wants to mount Python directories and also add Python scripts as `sgx.trusted_files` (see https://github.com/oscarlab/graphene/blob/master/Examples/python/python.manifest.template).

As for Python installation paths:
- Ubuntu/Debian are special and use non-standard `/usr/lib/python3/dist-packages` directory to store all user-installed packages.
- Fedora seems to use `/usr/lib64/python3.6/site-packages`.
- CentOS 8 uses `/usr/local/lib64/python3.6/site-packages`.

If we want to have One Manifest to Rule Them All, then we need to somehow mount one of the above directories, depending on the distro. Python has two packages to help with identifying these directories: `sysconfig` and `distutils.sysconfig`.
Woju currently set up Graphene such that it takes `sysconfig.get_path('stdlib')` and modifies it to get to the final expected Ubuntu path.

The closest I could get to the above three options is using `distutils.sysconfig.get_python_lib(plat_specific=True)` (correct path on Ubuntu, correct path on Fedora, but *wrong* path on CentOS)...

It gets even worse: ideally, we want to mount not just one Python path but a couple of them (e.g. `/usr/lib64/python3.6/site-packages` and `/usr/lib/python3.6/site-packages` -- note lib64 vs lib). This "one or several Python-path mount points" is not supported in our current manifest at all.

At this point, it looks like the only reasonable option is to force users to amend our Ubuntu-specific example manifest for their target platform. In other words, users must modify our Ubuntu-specific `python.manifest.template` to their own `python-centos.manifest.template`.

**Comment by dimakuv** [comment]
For Issue 1, I submitted https://github.com/oscarlab/graphene/pull/2509 (adds a note to our documentation).

Issue 2 looks unsolvable to me. One has to adopt a manifest file for each new target OS distro (or use venv of Python, but this again means writing a new manifest file).

Finally, Issue 3 (`/etc/default/apport`) can be easily circumvented by removing this line from the manifest file. Since Apport is Ubuntu-specific, this file won't be needed on CentOS.

@anjalirx-intel These are the solutions I propose to your issue. In my opinion, we should have separate `manifest.template` files for CentOS/Fedora, and not try to "stretch" our existing manifest files for all OS distros. CentOS/Fedora is just way to different from Ubuntu/Debian...

**Comment by mkow** [comment]
@anjalirx-intel ping ^
I think we can close this issue now?

**Comment by anjalirx-intel** [comment]
@mkow Yes, we can. Closing the issue now
